### PR TITLE
Add online newton optimizer

### PR DESCRIPTION
**Pull request by eserie**

### CODE REVIEW

**Comment by mkunesch** [comment]
Hi! Thanks a lot for the PR! Could you fix the pylint errors before we review (so that we can focus on the important things in the review)? Thanks a lot!

**Comment by eserie** [comment]
Hi, sure. I fixed the pylint error in an additional commit. Feel free to ask me to change things. 
For now, the dev is more or less as I originally developed it and the test coverage is certainly a bit too limited.

**Comment by mkunesch** [comment]
Hi! Thanks for fixing the pylint error - could you also do the same in the test file? I think that's still holding up the checks.
Due to the holidays it might take us slightly longer to review so there is no rush.

**Comment by eserie** [comment]
Hi! I fixed the tests! Normally all checks should pass now.

**Comment by mkunesch** [comment]
Great, thank you very much for fixing all the checks! I've assigned myself and should have time to review this week.
Thanks a lot again!

File: optax/_src/alias_test.py
Comment (mkunesch) [inline optax/_src/alias_test.py:128]: Nit: I know most tests in this file don't do this, but in theory test names and/or docstrings should explain what the test checks. An example here could be `test_online_newton_step_decreases_loss(self)`, with a similar single-line docstring. This will be shown in the error message from the test when the tests fails and is therefore very helpful when debugging.

File: optax/_src/transform.py
Comment (mkunesch) [inline optax/_src/transform.py:874]: Nit: I think we don't have this whitespace for other states.

File: optax/_src/transform.py
Comment (mkunesch) [inline optax/_src/transform.py:878]: I think I'd expect this to be a private function or is there a use-case of `sherman_morrison` directly which we would like to support?

Comment (mkunesch) [inline optax/_src/transform.py:878]: Please add type annotations to this function (even if we decide to make it private in the end).

File: optax/_src/transform.py
Comment (mkunesch) [inline optax/_src/transform.py:885]: Please shorten this line so that the short description fits onto a single line (with more detail in the detailed docstring if necessary).

File: optax/_src/transform.py
Comment (mkunesch) [inline optax/_src/transform.py:888]: As above, please format this similar to e.g. `rmsprop` in `alias.py` disabling and re-enabling the `line-too-long` warning if necessary.

File: optax/_src/transform.py
Comment (mkunesch) [inline optax/_src/transform.py:910]: Do I understand correctly that you wouldn't need this if you don't overwrite `updates` in line 917 and instead introduce a new variable e.g. `flattened_updates` and then in line 922:

```
updates = jax.tree_multimap(
  lambda flat_u, u: flat_u.reshape(u.shape), flattened_updates, updates)
```

If that's true I think it would be good to simplify the code and remove the definition the `Tuple`.

File: optax/_src/transform_test.py
Comment (mkunesch) [inline optax/_src/transform_test.py:213]: Please use the corresponding unittest/absltest/np.testing asserts so that the error messages from the test are more useful - in this case I think it would be `np.testing.assert_allclose`.

File: optax/_src/transform_test.py
Comment (mkunesch) [inline optax/_src/transform_test.py:207]: What does the `* 1.0` do?

File: optax/_src/transform_test.py
Comment (mkunesch) [inline optax/_src/transform_test.py:206]: This is a very special value: in line 210 it will mean that `x` doesn't have an effect. Could you set up the test such that `x` is used in a non-trivial way?

File: optax/_src/transform_test.py
Comment (mkunesch) [inline optax/_src/transform_test.py:208]: All of these test values look like special cases to me. Could you add a test with more general conditions? (If the special case is important to test please keep it around as a separate test.)

File: optax/_src/transform_test.py
Comment (mkunesch) [inline optax/_src/transform_test.py:205]: If we make the function `sherman_morrison` private, it shouldn't be tested here, but instead will be tested indirectly through the test of the public `scale_by_online_newton`.

Comment (mkunesch) [reply optax/_src/transform_test.py:205]: typo: extra s in morrison.

File: optax/_src/transform_test.py
Comment (mkunesch) [inline optax/_src/transform_test.py:214]: Please add a test for `scale_by_newton` that tests both `init` and `update`. The update function can be tested e.g. by comparing an analytic calculation against a step taken with the code.

File: optax/_src/transform.py
Comment (mkunesch) [inline optax/_src/transform.py:892]: Could you add which algorithm/formula/section it is in the paper so that it's easier for the reader to cross-reference?

File: optax/_src/transform.py
Comment (mkunesch) [inline optax/_src/transform.py:900]: Note: I haven't reviewed the rest of this file yet as I'm not sure which equations exactly they implement in the paper. Please point me to the algorithm, equation, or section. Thanks a lot!

**Comment by mkunesch** [comment]
Hi! Thank you so much for the contribution again!

I have been reviewing your code, but in order to check the equations it would be very helpful if you could point me to exactly the equation, algorithm, or section in the paper that you have implemented so that I can compare it to the code more easily.

I will add the comments I have so far below so that they don't get lost in the meantime, but please don't make the changes yet as I'm still reviewing the code.

Thanks a lot for your help and for this PR!

**Review by mkunesch — CHANGES_REQUESTED** [review]
Hi! Thank you so much for the contribution again!

I have been reviewing your code, but in order to check the equations it would be very helpful if you could point me to exactly the equation, algorithm, or section in the paper that you have implemented so that I can compare it to the code more easily.

I will add the comments I have so far below so that they don't get lost in the meantime, but please don't make any changes yet as I'm still reviewing the code.

Thanks a lot for your help and also thanks again for this PR!

File: optax/_src/transform.py
Comment (eserie) [reply optax/_src/transform.py:900]: Hi @mkunesch, thanks for your review! 
For your last question, the algorithm implemented is the ONS (ONLINE NEWTON STEP) Fig.2 in the paper.

Comment (mkunesch) [reply optax/_src/transform.py:900]: Amazing, thanks a lot - now I understand! I have a few follow-up questions but I'll ask those for specific lines.
(It might be worth adding this to the doc string just because there is additional boilerplate in the paper and it takes a while to see why only the terms in the code remain - thanks a lot!)

File: optax/_src/transform.py
Comment (mkunesch) [inline optax/_src/transform.py:903]: I have never implemented these sherman-morrison updates so I don't know whether this is actually an issue, but from looking at the code, I'd be worried about catastrophic cancellation when starting with jnp.eye / eps which has diagonal entries of O(10^7) and doing lots of additions and subtractions in the first step.

Is this something you have thought about? Is this a problem here? Perhaps there is a way to take the limit eps -> 0 explicitly for the first step to avoid this?

Thanks a lot!

File: optax/_src/transform.py
Comment (mkunesch) [reply optax/_src/transform.py:878]: If we decide to make this private, I'd also change the implementation to specific to our use case where u == v. In any case, please also add a docstring explaining what this does and what sherman-morrison is used for in this case. Thanks a lot!

File: optax/__init__.py
Comment (eserie) [reply optax/__init__.py:32]: Ok, I will correct it (along with the change of name to online_newton)!
(Suggestion: using isort could permit to manage this automatically.)

File: optax/_src/alias.py
Comment (eserie) [reply optax/_src/alias.py:676]: This is for pep8 formatting. What is the number of characters per line you accept in the project?
(I used "black" which defaults to 88 characters per line).

File: optax/_src/transform.py
Comment (eserie) [reply optax/_src/transform.py:892]: Yes the correct link is this one: https://link.springer.com/content/pdf/10.1007/s10994-007-5016-8.pdf

File: optax/_src/transform.py
Comment (eserie) [reply optax/_src/transform.py:903]: You are absolutely right.  In fact, in practice, this parameter may not be small. It should be chosen with some apriori knowledge about the scale of parameters and gradients (e.g. by following the formulas in the article).
Perhaps we should not define a default value for this parameter to let the user choose it?

**Comment by eserie** [comment]
Hi @mkunesch, thanks for your review.
I tried to address your remarks and also made a rebase against the master branch.
Let me know if anything is wrong.

File: optax/_src/transform.py
Comment (mkunesch) [reply optax/_src/transform.py:944]: Okay, yes, letting the user choose it is helpful - do you have an intuition for how generally applicable the default value would be? If it always has to be tuned, I'd remove the default entirely and force the user to choose it.

Also, is there an explanation (or a paper or reference implementation) we can give the user to help them choose `eps`? I don't think the paper presents experiments with this, does it?

**Comment by mkunesch** [comment]
Hi @eserie!

How you are getting on with the other changes? No rush at all, but I wanted to make sure you are not waiting for my review of the changes you have already made in the meantime! Thanks a lot for making them - I've added one comment on the change re `eps`.

Thanks a lot!

**Comment by eserie** [comment]
Hi @mkunesch !
I think I reviewed all of your comments and I should have addressed them. Can you tell me if sounds good to you?

File: optax/_src/alias_test.py
Comment (mkunesch) [reply optax/_src/alias_test.py:149]: @eserie

File: optax/_src/transform_test.py
Comment (mkunesch) [reply optax/_src/transform_test.py:249]: @eserie

File: optax/_src/transform.py
Comment (mkunesch) [reply optax/_src/transform.py:950]: @eserie

File: optax/_src/transform_test.py
Comment (mkunesch) [reply optax/_src/transform_test.py]: @eserie

File: optax/_src/transform.py
Comment (mkunesch) [reply optax/_src/transform.py:943]: @eserie

**Comment by mkunesch** [comment]
Hi! 
Ah, I think GitHub collapsed these comments in the discussion view. There were a few more - I've tagged in you in some (but not all) of them so that GitHub shows them for you. They should also be visible in the _Files Changed_ view.

Some of them are with regards to the tests - but we could also merge into experimental where we require less testing for new code.

**Comment by mtthss** [comment]
Hello @eserie, can we get this submitted?

**Comment by eserie** [comment]
Hello @mtthss, @mkunesch , sorry I've been pretty busy the last 3 months. @mkunesch, I can see your comments in  Files Changed view, sorry I missed them before. Hopefully, I can take a look at it by the end of the month, if that's ok with you.

File: optax/_src/transform.py
Comment (eserie) [reply optax/_src/transform.py]: Thank you for the observation! I was a bit messy in my implementation, your proposal seems much more reasonable. I updated the code accordingly and also add a test to cover the multi-dimensional case.

**Comment by eserie** [comment]
Hi, I finally get to run over all your comment and added a new test for the multi-dimensional weights case. 
I finally did a rebase of my changes on the current master branch and fuse all my commits in one.
I also ran the tests and checks as in the test.sh file, everithing looks good.

**Comment by mtthss** [comment]
There is still one change requested by @mkunesch, could you address and then we get this submitted? thanks a lot for the contribution!

**Comment by mtthss** [comment]
This branch has conflicts that must be resolved

**Comment by mkunesch** [comment]
Hi! Thanks a lot @eserie for making the changes. I just returned from holiday and will make sure to review by the end of the week.

**Comment by eserie** [comment]
Hi @mkunesch, thank you! Let me know if everything looks good to you. Unfortunatly, I accidentally closed the pull request while rebasing my branch. Can you open it again? The last commit to reattach is [fa028f5](https://github.com/eserie/optax/commit/fa028f572e4632a0841876e9333e3ed0782f843d).

File: optax/_src/alias.py
Comment (mkunesch) [inline optax/_src/alias.py:436]: Actually, I think as long as the link can fit into a line we could just introduce line breaks in the references. At least in the new theme e.g. for [rmsprop here](https://optax.readthedocs.io/en/latest/api.html#optax.rmsprop) I don't think there is a benefit to disabling and having it in one line.

File: optax/_src/alias.py
Comment (mkunesch) [inline optax/_src/alias.py:450]: (If we do end up keeping the long line we should enable the warning again immediately after the docstring.)

File: optax/_src/alias.py
Comment (mkunesch) [inline optax/_src/alias.py:438]: According to the google python style guide there needs to be a blank line following the summary line.

File: optax/_src/alias_test.py
Comment (mkunesch) [inline optax/_src/alias_test.py:146]: typo: ... newton step update decreases a

But actually, the "We" isn't typical for test docstrings. I'd suggest:

```
"""Checks that the online newton step update decreases the loss."""
```

File: optax/_src/transform.py
Comment (mkunesch) [inline optax/_src/transform.py:962]: The first line of the docstring should be a <=80 character summary according to https://google.github.io/styleguide/pyguide.html#381-docstrings and followed by a blank line.

In this case you could just keep Sherman Morrison formula as summary and move the rest to the detailed description.

Please also add an Args and Returns section as the internal linter won't allows us to merge non-trivial docstrings (i.e. docstrings that don't just say "Returns X given Y.") without Args and Returns sections.

File: optax/_src/transform.py
Comment (mkunesch) [inline optax/_src/transform.py:961]: Could you add type annotations to this function please? I assume, here this is just `chex.Array` for all inputs and the return type?

(Not all of optax was typed in the past but we'd like all new contributions to use type annotations and are slowly adding them to the old code too.)

File: optax/_src/transform.py
Comment (mkunesch) [inline optax/_src/transform.py:967]: Nit: there is enough space to write out den to denominator so I'd do this to make the code easier to read.

File: optax/_src/transform.py
Comment (mkunesch) [inline optax/_src/transform.py:985]: (As above, we could actually break the lines in the references I think but in any case we should re-enable the line-too-long warning again here otherwise the rest of the function doesn't get checked.)

File: optax/_src/transform.py
Comment (mkunesch) [inline optax/_src/transform.py:975]: As above, the docstrings should have a <= 80 character summary line followed by a blank line.

File: optax/_src/transform.py
Comment (mkunesch) [inline optax/_src/transform.py:996]: `Tree_multimap` has been deprecated by jax. You can use `tree_map` as a drop-in replacement.

File: optax/_src/transform.py
Comment (mkunesch) [inline optax/_src/transform.py:999]: Nit: I think `h_inv` would be easier to read.

File: optax/_src/transform_test.py
Comment (mkunesch) [inline optax/_src/transform_test.py:281]: Nit: I'd write out ref to `correct_inverse` to make the code and the errors that are raised if the test fails more readable.

File: optax/_src/transform_test.py
Comment (mkunesch) [inline optax/_src/transform_test.py:310]: I think these three lines of the test can be removed for a unit test. The last part of the test is great: it checks the calculated gradients against the analytic calculation that we promise to the user. Whether or not sherman_morrison gets called is more of an implementation detail (indeed I think it could be a private function) and testing that isn't typical for a unit test.

File: optax/_src/transform_test.py
Comment (mkunesch) [inline optax/_src/transform_test.py:340]: (As above, I'd remove these three lines - the important part is the comparison agains the analytical results).

File: optax/_src/transform_test.py
Comment (mkunesch) [inline optax/_src/transform_test.py:305]: Could you wrap `step.update` and `step.init` in `self.variant` and decorate the test with `@chex.all_variants` please? You can look at some of the tests at the top of the file for examples.

This ensures that the optimizer code can be jit compiled, executed on device etc.

**Review by mkunesch — CHANGES_REQUESTED** [review]
Hi! Thanks a lot for making all the changes! 

I did a second pass now with more detailed comments (mostly formatting).

The checks currently fail - but I think that's due to `tree_multimap` being deprecated. You can just replace it by `tree_map` and it should work.

Thanks a lot again!

**Comment by eserie** [comment]
Hi @mkunesch, thanks for all your comments. I tried to address them all, hope all is ok now!

**Comment by mkunesch** [comment]
Hi!

Thanks a lot! Before merging I wanted to make sure I could use it myself, but had difficulty finding parameters that could find the minimum of a parabola.

Would it be possible to add the optimizer to the test `test_optimization` in `alias_test.py` that runs all the optimizers on a parabola and ensures they find the correct minimum? That way there would be examples of the parameters that work.

Other than that, the code looks good to me up to minor formatting edits and capitalization etc that I could fix with the internal linter during merging if that's ok with you! (But if you prefer I'm also happy to make file comments and you can fix them).

Thanks again!

**Comment by eserie** [comment]
Hi!

I could add the `online_newton_step` optimizer to the test `test_optimization` in `alias_test.py`. However, I could not find parameters that fits for the two loss functions. So I addded a special condition in test code to adjust the parameters for one of the loss functions. I hope this is ok.

No problem for me if you want to fix the formatting issues you have seen (on my side the checks done in the test.sh all passed).

NB: to find the parameters, I played with this code in an interactive session:
```python
import jax
import jax.numpy as jnp
import optax
from matplotlib import pyplot as plt
from optax._src import alias
from optax._src import update
from optax._src import numerics

a = 1.0
b = 100.0

initial_params = jnp.array([0.0, 0.0])
final_params = jnp.array([a, a**2])


def fun(params):
    return numerics.abs_sq(a - params[0]) + b * numerics.abs_sq(
        params[1] - params[0] ** 2
    )

opt = alias.online_newton_step(5.0e-1, eps=1.)

params = initial_params
state = opt.init(params)


@jax.jit
def step_(state_params, i):
    state, params = state_params
    val, grad = jax.value_and_grad(fun)(params)
    updates, state = opt.update(grad, state)
    params = update.apply_updates(params, updates)
    return (state, params), (params)


state, params_history = jax.lax.scan(step_, (state, params), jnp.arange(10000))
plt.plot(jnp.linalg.norm(params_history - final_params[None, :], axis=1))
plt.title("$| w - w^* |$")
jnp.linalg.norm(params_history[-1] - final_params)
```
image

**Comment by mkunesch** [comment]
Hi @eserie,

Thanks a lot for adding the optimizer to the tests and finding parameter combinations that work!

We have experimented more with the optimizer in the past month (trying it on various functions and some deep learning work loads such as MNIST) and we have concluded that this optimizer may not be a good fit for the core optax API at this moment.

Optax is currently focused on optimizers that can be substituted into most deep learning training loops (ideally with the default parameters), and we have found that when using ONS on DL problems, finding the right parameters can be tricky and it may run out of memory for typical DL networks.

From the paper and what you have written, ONS is at its best in online learning on time series and streaming data, so we would suggest publishing this optimizer as part of a repository that specializes in these applications. You can of course still use the optax machinery and make this work with optax, but we would suggest providing this as a third-party component that uses optax rather than integrating it into the core API. We would of course be happy to prominently post a link to it here in this PR (or a new issue) so that people can find it easily if they search for it in optax.

Sorry to have decided this towards the end of the code review. We will try to prevent this in the future by creating a series of problems every optimizer should solve with standard parameters before starting the review. Still, thank you very much for making all the changes and we hope that the comments are useful for you for your own open-source version!

Thanks a lot for filing the PR again and we hope you understand the decision!

**Pull request closed without merging**
